01_P/P_2_2_2_02/sketch.js

```diff
@@ -0,0 +1,142 @@
+// limitations under the License.
+
+/**
+ * draw the path of an intelligent agent
+ *
+ * MOUSE
+ * position x          : composition speed of the picture
+ *
+ * KEYS
+ * Delete/Backspace    : clear display
+ * 1-3                 : draw mode
+ * s                   : save png
+ */
+'use strict';
+
+var NORTH = 0;
+var EAST = 1;
+var SOUTH = 2;
+var WEST = 3;
+var direction = SOUTH;
+
+var stepSize = 3;
+var minLength = 10;
+var diameter = 1;
+var angleCount = 7;
+var angle;
+var reachedBorder = false;
+
+var posX;
+var posY;
+var posXcross;
+var posYcross;
+
+var dWeight = 50;
+var dStroke = 4;
+
+var drawMode = 1;
+
+function setup() {
+    createCanvas(600, 600);
+    colorMode(HSB, 360, 100, 100);
+    background(360, 0, 100);
+
+    angle = getRandomAngle(direction);
+    posX = floor(random(width));
+    posY = 5;
+    posXcross = posX;
+    posYcross = posY;
+}
+
+function draw() {
+    // var speed = int(map(mouseX, 0, width, 0, 20));
+
+    for (var i = 0; i <= mouseX; i++) {
+
+        // ------ draw dot at current position ------
+        strokeWeight(1);
+        stroke(180, 0, 0);
+        point(posX, posY);
+
+        // ------ make step ------
+        posX += cos(radians(angle)) * stepSize;
+        posY += sin(radians(angle)) * stepSize;
+
+        // ------ check if agent is near one of the display borders ------
+        reachedBorder = false;
+
+        if (posY <= 5) {
+            direction = SOUTH;
+            reachedBorder = true;
+        } else if (posX >= width - 5) {
+            direction = WEST;
+            reachedBorder = true;
+        } else if (posY >= height - 5) {
+            direction = NORTH;
+            reachedBorder = true;
+        } else if (posX <= 5) {
+            direction = EAST;
+            reachedBorder = true;
+        }
+
+        // ------ if agent is crossing his path or border was reached ------
+        loadPixels();
+        var currentPixel = get(floor(posX), floor(posY));
+        if (
+            (
+                currentPixel[0] === 0 &&
+                currentPixel[1] === 0 &&
+                currentPixel[2] === 0
+            ) ||
+            reachedBorder
+        ) {
+            angle = getRandomAngle(direction);
+            var distance = dist(posX, posY, posXcross, posYcross);
+            if (distance >= minLength) {
+                strokeWeight(distance / dWeight);
+                if (drawMode == 1) stroke(0);
+                if (drawMode == 2) stroke(52, 100, distance / dStroke);
+                if (drawMode == 3) stroke(192, 100, 64, distance / dStroke);
+
+                line(posX, posY, posXcross, posYcross);
+            }
+            posXcross = posX;
+            posYcross = posY;
+        }
+    }
+}
+
+function keyReleased() {
+    if (key == 's' || key == 'S') saveCanvas(gd.timestamp(), 'png');
+    if (keyCode === DELETE || keyCode === BACKSPACE) background(360);
+    if (key === ' ') noLoop();
```

Comment:
Ah! Good catch. I should have removed this. I will remove line 128. Thanks for pointing this out.